NODE-661 DAG Storage RDBMS

igor-ramazanov opened the pull request:
### Overview
Adds new `SQLiteDagStorage` and replaces `FileDagStorage`.

### Which JIRA ticket does this PR relate to?
https://casperlabs.atlassian.net/browse/NODE-661

### Complete this checklist before you submit this PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [CasperLabs coding standards](https://casperlabs.atlassian.net/wiki/spaces/EN/pages/16842753/Coding+Standards).
- [x] If this PR adds a new feature, it includes tests related to this feature.
- [x] You assigned one person to review this PR.
- [x] Your GitHub account is linked with our [Drone CI](http://drone.casperlabs.io/) system. This is necessary to run tests on this PR.

### Notes
1. This PR will be merged to the https://github.com/CasperLabs/CasperLabs/tree/NODE-659-rdbms-epic instead of `dev`.
2. Pay attention that even though this PR doesn't add any new special tests for `DagStorage` it's still covered with tests of the `casper` module. Please, don't hesitate if you still think I should write laws-like (similar to `DeployStorage`) tests for `DagStorage` trait regardless of implementation.

goral09 commented on storage/src/main/resources/db/migration/V20190820_661__Create_dag_storage_tables.sql:
```diff
@@ -0,0 +1,26 @@
+
+CREATE TABLE dag_storage_blocks_metadata
+(
+    block_hash BLOB NOT NULL PRIMARY KEY,
```
I would store creator hash and rank somewhere. We often do queries like "the last message by validator".

igor-ramazanov replied:
Done

goral09 commented on storage/src/main/resources/db/migration/V20190820_661__Create_dag_storage_tables.sql:
```diff
@@ -0,0 +1,26 @@
+    PRIMARY KEY (parent_block_hash, child_block_hash, parent_rank)
+) WITHOUT ROWID;
+
+CREATE TABLE dag_storage_validators_latest_messages
```
Does it have to be a separate table? It will constantly be getting updated (basically on every new block). If there was a table with a primary key on `(validator, rank)` then we could do queries that group by `validator`, sort by `rank` and pick `limit 1`. Which effectively is `validators_latest_messages`.

igor-ramazanov replied:
Done

goral09 commented on storage/src/main/scala/io/casperlabs/storage/dag/SQLiteDagStorage.scala:
```diff
@@ -0,0 +1,280 @@
+    sql"""|SELECT validator, data
+          |FROM (SELECT validator, data, MAX(rank)
+          |      FROM dag_storage_blocks_metadata
+          |      WHERE validator!=${ByteString.EMPTY}
```
Is that b/c of the Genesis not having a validator? Won't it be better, from the performance pov, to filter that entry (there should be only one) in the application layer instead of having to do that comparison in DB for every row?

igor-ramazanov replied:
Done

goral09 commented on storage/src/main/scala/io/casperlabs/storage/dag/SQLiteDagStorage.scala:
```diff
@@ -0,0 +1,280 @@
+}
+
+object SQLiteDagStorage {
+  def create[F[_]: Sync](implicit xa: Transactor[F], met: Metrics[F]): F[DagStorage[F]] =
```
Should this method be private?

igor-ramazanov replied:
Done

goral09 commented on storage/src/main/scala/io/casperlabs/storage/dag/SQLiteDagStorage.scala:
```diff
@@ -0,0 +1,280 @@
+      .transact(xa)
+      .flatMap(rearrangeSQLiteResult)
+
+  override def topoSort(startBlockNumber: Long): F[Vector[Vector[BlockHash]]] =
```
I feel like this could be a stream. I can imagine `startBlockNumber=0`. We don't want to pull into the application layer all of the results at once.

igor-ramazanov replied:
I agree that we'd like to stream results instead of allocating all of them at once. 

But if I change these methods to streaming then it will lead to quite a lot of changes, because I'll need to change the node's gRPC layer, since these methods used to send data to node's API clients.

I'm okay working on it in the scope of this PR if you think that's okay. Or I can create a ticket and work on it in the future.

goral09 replied:
in the gRPC you could decide to drain them all right away (like `.toList` or sth) but option would be there.

igor-ramazanov replied:
Done, fs2 has an awesome API which allows closing the RDBMS results set early.

https://gitter.im/tpolecat/doobie?at=5d5fdccb07d1ff39f86d1d44

goral09 commented on storage/src/main/resources/db/migration/V20190820_661__Create_dag_storage_tables.sql:
```diff
@@ -0,0 +1,28 @@
+);
+
+CREATE INDEX idx_dag_storage_blocks_metadata_validator_rank
+    ON dag_storage_blocks_metadata (validator, rank);
```
Maybe I was wrong about this index. It looks like `dag_storage_blocks_metadata` table is always either grouped by or searched for the specific validator. Index on `validator` should be enough in this case.

igor-ramazanov replied:
I reverted back a separate table for latest messages.

goral09 commented on storage/src/test/scala/io/casperlabs/storage/dag/DagStorageTest.scala:
```diff
@@ -479,3 +484,54 @@ class FileDagStorageTest extends DagStorageTest {
+        .dataSource(s"jdbc:sqlite:$db", "", "")
+        .locations(new Location("classpath:db/migration"))
+    conf.load()
+  }
```
A lot of this setup work looks like sth in `SQLiteDeployStorageSpec`.

igor-ramazanov replied:
Done

goral09 commented on casper/src/test/scala/io/casperlabs/casper/util/comm/CasperPacketHandlerSpec.scala:
```diff
@@ -114,10 +121,8 @@ class CasperPacketHandlerSpec extends WordSpec with Matchers {
-      .create[Task]
-      .unsafeRunSync(monix.execution.Scheduler.Implicits.global)
-    implicit val casperRef = MultiParentCasperRef.unsafe[Task](None)
+    implicit val inMemDagStorage  = createSQLiteDagStorage()
```
It says `inMemDagStorage` but it's now replaced with SQLite one.

igor-ramazanov replied:
Done

goral09 commented on casper/src/test/scala/io/casperlabs/casper/util/comm/CasperPacketHandlerSpec.scala:
```diff
@@ -129,6 +134,32 @@ class CasperPacketHandlerSpec extends WordSpec with Matchers {
     implicit val validation = new ValidationImpl[Task]
   }
 
+  private def createSQLiteDagStorage()(implicit metrics: Metrics[Task]) = {
```
Could you reuse `DagStorageFixture.createDagStorage` here?

igor-ramazanov replied:
Done

aakoshh commented on storage/src/main/scala/io/casperlabs/storage/dag/SQLiteDagStorage.scala:
```diff
@@ -0,0 +1,280 @@
+          |FROM (SELECT block_hash, MAX(rank)
+          |      FROM dag_storage_blocks_metadata
+          |      WHERE validator = $validator
+          |      GROUP BY validator)""".stripMargin
```
This doesn't look like valid SQL, that the `block_hash` is in the projection without an aggregation function, or being in the group by as well. Is it some SQLite specific thing? Oracle has a [keep](https://docs.oracle.com/cd/B19306_01/server.102/b14200/functions056.htm) feature.

Plus because the `validator` can only have 1 value, there's no reason for it to be in the group by.

I still think a separate table would be faster.

igor-ramazanov replied:
Done

aakoshh commented on storage/src/test/resources/logback-test.xml:
```diff
@@ -0,0 +1,19 @@
+<configuration>
+
```
Shouldn't this file be named `logback.xml`?

igor-ramazanov replied:
As I understand it's preferable to use `logback-test.xml` in test resource: https://stackoverflow.com/a/50579418/7688649

aakoshh replied:
Ah, cool, I didn't know it looked for this, thanks.

goral09 commented on storage/src/main/scala/io/casperlabs/storage/dag/SQLiteDagStorage.scala:
```diff
@@ -0,0 +1,280 @@
+           |""".stripMargin
+      ).updateMany(
+        toUpdateValidators
+          .map((blockMetadata.blockHash, _, blockMetadata.rank, blockMetadata.toByteString))
```
I suggested that change to Igor (he had it separate at the beginning). I thought that it will be more difficult to keep tables consistent (transaction would have to span more tables) and having an index on `(validator, rank)` would be enough to efficiently retrieve the latest message by a validator. Think about it now, you might be right that we always want to retrieve _the last message by validator_ so having an index on _ALL_ of its messages is an overkill.

Sorry for confusing you @igor-ramazanov . You were right at the beginning.

igor-ramazanov replied:
Done

igor-ramazanov commented on storage/src/main/scala/io/casperlabs/storage/dag/SQLiteDagStorage.scala:
```diff
@@ -0,0 +1,280 @@
+      .map { blockHashes =>
+        val order = blockHashes.zipWithIndex.toMap
+        Ordering.by(blockMetadata => order(blockMetadata.blockHash))
+      }
```
Done

igor-ramazanov commented on storage/src/main/scala/io/casperlabs/storage/dag/SQLiteDagStorage.scala:
```diff
@@ -0,0 +1,280 @@
+      if (block.isGenesisLike) {
+        sql"""|INSERT OR IGNORE INTO dag_storage_topological_sorting 
+              |(parent_block_hash, child_block_hash, child_rank) 
+              |VALUES (${ByteString.EMPTY}, ${blockMetadata.blockHash}, 0)""".stripMargin.update.run
```
Done

igor-ramazanov commented on storage/src/main/resources/db/migration/V20190820_661__Create_dag_storage_tables.sql:
```diff
@@ -0,0 +1,28 @@
+CREATE TABLE dag_storage_topological_sorting
```
Done

igor-ramazanov commented on storage/src/main/resources/db/migration/V20190820_661__Create_dag_storage_tables.sql:
```diff
@@ -0,0 +1,28 @@
+    PRIMARY KEY (justification_block_hash, block_hash)
+) WITHOUT ROWID;
+
+CREATE TABLE dag_storage_blocks_metadata
```
Done

igor-ramazanov commented on storage/src/main/scala/io/casperlabs/storage/dag/SQLiteDagStorage.scala:
```diff
@@ -0,0 +1,280 @@
+  import io.casperlabs.models.BlockImplicits._
+
+  implicit val metaByteString: Meta[ByteString] =
+    Meta[Array[Byte]].imap(ByteString.copyFrom)(_.toByteArray)
```
Done

igor-ramazanov commented on storage/src/main/resources/db/migration/V20190820_661__Create_dag_storage_tables.sql:
```diff
@@ -0,0 +1,28 @@
+(
+    parent_block_hash BLOB    NOT NULL,
+    child_block_hash  BLOB    NOT NULL,
+    child_rank        INTEGER NOT NULL,
```
Done

igor-ramazanov commented on storage/src/main/resources/db/migration/V20190820_661__Create_dag_storage_tables.sql:
```diff
@@ -0,0 +1,28 @@
+    block_hash BLOB    NOT NULL PRIMARY KEY,
+    validator  BLOB    NOT NULL,
+    rank       INTEGER NOT NULL,
+    data       BLOB    NOT NULL
```
Done

igor-ramazanov commented on storage/src/main/resources/db/migration/V20190820_661__Create_dag_storage_tables.sql:
```diff
@@ -0,0 +1,28 @@
+(
+    block_hash BLOB    NOT NULL PRIMARY KEY,
+    validator  BLOB    NOT NULL,
+    rank       INTEGER NOT NULL,
```
Done

igor-ramazanov commented on storage/src/main/resources/db/migration/V20190820_661__Create_dag_storage_tables.sql:
```diff
@@ -0,0 +1,28 @@
+    parent_block_hash BLOB    NOT NULL,
+    child_block_hash  BLOB    NOT NULL,
+    child_rank        INTEGER NOT NULL,
+    PRIMARY KEY (parent_block_hash, child_block_hash, child_rank)
```
Done

igor-ramazanov commented on casper/src/test/scala/io/casperlabs/casper/helper/DagStorageFixture.scala:
```diff
@@ -102,12 +102,32 @@ object DagStorageTestFixture {
+  def createDagStorage[F[_]: Metrics: Async: ContextShift](
+      dagStorageDir: Path,
+      maybeGenesis: Option[Block] = None
+  )(): F[DagStorage[F]] = {
```
Forgot to remove after removing implicit parameters. Done

igor-ramazanov commented on casper/src/main/scala/io/casperlabs/casper/validation/ValidationImpl.scala:
```diff
@@ -398,6 +385,7 @@ class ValidationImpl[F[_]: MonadThrowable: FunctorRaise[?[_], InvalidBlock]: Log
                 s"block number $actuallyRank is not the maximum block number of justifications plus 1, i.e. $calculatedRank."
             for {
               _ <- Log[F].warn(ignore(b, logMessage))
+              _ = println(logMessage)
```
Done

igor-ramazanov commented on node/src/main/resources/default-configuration.toml:
```diff
@@ -127,9 +127,6 @@ use-tls = false
 [blockstorage]
 
-# Size factor for squashing block storage latest messages.
-latest-messages-log-max-size-factor = 10
```
It was used only by `FileDagStorage` in the `NodeRuntime` which is replaced to the `SQLiteDagStorage`. Do you think it should be added back?

aakoshh replied:
No, it's fine, I just wasn't sure if you are already chucking that one out or it was never used.

igor-ramazanov commented on casper/src/test/scala/io/casperlabs/casper/helper/DagStorageFixture.scala:
```diff
@@ -102,12 +102,32 @@ object DagStorageTestFixture {
+      dagStorageDir: Path,
+      maybeGenesis: Option[Block] = None
+  )(): F[DagStorage[F]] = {
+    val jdbcUrl = s"jdbc:sqlite:${dagStorageDir.resolve("dag_storage.db")}"
```
Left a TODO here for the future

igor-ramazanov commented:
@goral09 @aakoshh 
Thank you for the thorough reviews! You found many issues that otherwise could go to the main branch.

 I've addressed your comments, please take a look again.

goral09 commented on casper/src/main/scala/io/casperlabs/casper/finality/singlesweep/FinalityDetectorBySingleSweep.scala:
```diff
@@ -87,7 +88,7 @@ class FinalityDetectorBySingleSweepImpl[F[_]: Monad: Log] extends FinalityDetect
       dag: DagRepresentation[F],
       committeeApproximation: Set[Validator],
-      levelZeroMsgs: Map[Validator, List[BlockMetadata]],
+      levelZeroMsgs: Map[Validator, List[BlockSummary]],
```
That change could have been done in separate PR. Diff caused by it is large.

aakoshh replied:
Wow, I didn't think you'd go straight to replace all of it! Thanks :)
Another alternative would have been to change the underlying type [here](https://github.com/CasperLabs/CasperLabs/blob/dev/storage/src/main/scala/io/casperlabs/blockstorage/BlockMetadata.scala#L17) but I think `BlockSummary` is quite ubiquitus at this point.

goral09 commented on casper/src/main/scala/io/casperlabs/casper/util/Sorting.scala:
```diff
@@ -1,11 +0,0 @@
-  implicit val byteArrayOrdering = Ordering.by(Base16.encode)
-
-  implicit val byteStringOrdering: Ordering[ByteString] = Ordering.by(_.toByteArray)
-}
```
I'm surprised it wasn't used. I am pretty sure I referenced it multiple times.

goral09 replied:
I saw that it was just moved.

goral09 commented on storage/src/main/scala/io/casperlabs/storage/BlockMetadata.scala:
```diff
@@ -1,58 +0,0 @@
-import io.casperlabs.casper.consensus.{Block, Bond}
-import scalapb.TypeMapper
-
-final case class BlockMetadata(
```
This thing was useful IMO. With protobuf you have to deal with all optionality and at some point call unsafe `.get` and pray it's not gonna throw.

goral09 commented on storage/src/main/scala/io/casperlabs/storage/dag/SQLiteDagStorage.scala:
```diff
@@ -0,0 +1,258 @@
+    with DagRepresentation[F]
+    with DoobieCodecs {
+  // Do not forget updating Flyway migration scripts at:
+  // resources/db/migrations
```
Who is this message to? :)

igor-ramazanov replied:
I thought it can be a useful reminder when someone makes changes in this class. But now I'm thinking that better would be relying on tests instead of such comments.

It came from a similar comment of the `SQLiteDeployStorage` where it appeared because previously Flyway scripts where located in the `block-storage` module while `SQLiteDeployStorage` at those time was in the `casper`, so it wasn't obvious where to find migration scripts.

Now, `SQLite...Storage` both located in the `storage` module together with migration scripts, so I think it's fine removing them. Done.

goral09 commented on storage/src/main/resources/db/migration/V20190820_661__Create_dag_storage_tables.sql:
```diff
@@ -0,0 +1,36 @@
+(
+    justification_block_hash BLOB NOT NULL,
+    block_hash               BLOB NOT NULL,
+    PRIMARY KEY (justification_block_hash, block_hash),
```
Does it make sense to add an index if these are the only two fields in the table? cc @aakoshh

goral09 replied:
I mixed primary key with an index.

aakoshh replied:
The primary key? Yes, as an integrity constraint. And if there were no indexes on it the table would still be slow, as it would just be a list of blocks without any particular order. If we add the `WITHOUT ROWID` then it essentially becomes a clustered index, stored as a B-Tree.

aakoshh commented on models/src/main/scala/io/casperlabs/models/BlockImplicits.scala:
```diff
@@ -0,0 +1,54 @@
+    def validatorBlockSeqNum: Int          = summary.getHeader.validatorBlockSeqNum
+    def validatorPublicKey: ByteString     = summary.getHeader.validatorPublicKey
+    def rank: Long                         = summary.getHeader.rank
+    def weightMap: Map[ByteString, Long] =
```
Thanks for these, @goral09 has been crying for it for 6 months :)

aakoshh commented on storage/src/main/resources/db/migration/V20190820_661__Create_dag_storage_tables.sql:
```diff
@@ -0,0 +1,36 @@
+    justification_block_hash BLOB NOT NULL,
+    block_hash               BLOB NOT NULL,
+    PRIMARY KEY (justification_block_hash, block_hash),
+    FOREIGN KEY (block_hash) REFERENCES block_metadata (block_hash)
```
`justification_block_hash` is a foreign key too.

igor-ramazanov replied:
Done

aakoshh commented on storage/src/main/resources/db/migration/V20190820_661__Create_dag_storage_tables.sql:
```diff
@@ -0,0 +1,36 @@
+) WITHOUT ROWID;
+
+CREATE UNIQUE INDEX idx_block_parents
+    ON block_parents (child_block_hash, parent_block_hash);
```
The `PRIMARY KEY` covers this, no need for a separate unique index.

aakoshh replied:
Oh, sorry, this is the other way around!

aakoshh commented on storage/src/main/scala/io/casperlabs/storage/util/DoobieCodecs.scala:
```diff
@@ -0,0 +1,32 @@
+import io.casperlabs.casper.consensus.Block.ProcessedDeploy
+import io.casperlabs.casper.consensus.{BlockSummary, Deploy}
+
+trait DoobieCodecs {
```
Nice, thanks!

aakoshh commented on storage/src/main/resources/db/migration/V20190820_661__Create_dag_storage_tables.sql:
```diff
@@ -0,0 +1,36 @@
+);
+
+CREATE UNIQUE INDEX idx_block_metadata_rank_block_hash
+    ON block_metadata (rank, block_hash);
```
This is trivially unique because `block_hash` has a primary key. I think you can remove the `UNIQUE` so SQLite doens't have to spend any time worrying about violating that rule.

goral09 replied:
Is the `(rank,  block_hash)` pair _trivially_ unique though?

aakoshh replied:
Well if the whole table can have only 1 record with a given block hash, then yes, that hash functionally determines the rank.

goral09 replied:
Of course.

igor-ramazanov replied:
Done

aakoshh commented on storage/src/main/resources/db/migration/V20190820_661__Create_dag_storage_tables.sql:
```diff
@@ -0,0 +1,36 @@
+
+CREATE TABLE block_metadata
+(
+    block_hash BLOB    NOT NULL PRIMARY KEY,
```
Would a `validator` key be useful here?

igor-ramazanov replied:
Yeah, it may be useful in the future, done.

aakoshh commented on storage/src/main/resources/db/migration/V20190820_661__Create_dag_storage_tables.sql:
```diff
@@ -0,0 +1,36 @@
+CREATE UNIQUE INDEX idx_block_parents
+    ON block_parents (child_block_hash, parent_block_hash);
+
+CREATE TABLE block_justifications
```
There should be an index similar to `idx_block_parents` on this one as well so we can quickly retrieve the justifications of a given block. Although right now we are not relying on DAG traversal based on this table.

goral09 replied:
Or we could take it from the `block_metadata` table.

aakoshh replied:
Yeah, that's where the program actually takes them from. `DagRepresentation` has `getChildren` but the opposite you are expected to achieve with `lookup`

igor-ramazanov replied:
Done

aakoshh commented:
Once this is merged it might be a good idea to pull `dev` into it. I used the BlockMetadata recently.

aakoshh commented:
Are we using fs2 because that's what comes with Doobie? Is there anything to know regarding the difference between fs2.Stream and monix Iterant?

igor-ramazanov commented:
@aakoshh Yes, doobie depends on fs2 internally. I don't know differences but I probably could expect that `Iterant` can be generally faster than `fs2.Stream`, because of experience of Monix maintainers in optimizing low-level stuff. API is mostly the same.
